### DISCUSSION TITLE
Where to find the source code for OOP TabControlDesigner?

### DISCUSSION THREAD

**Discussion by ulfemsoy** [dotnet/winforms#9564] [Q&A] [answered]
I would like to know more about how to create an out-of-process Control designer, and would like to see how the TabControl designer was implemented. However, I'm not able to find the TabControlDesigner class in the design assembly:
https://github.com/dotnet/winforms/tree/main/src/System.Windows.Forms.Design/src/System/Windows/Forms/Design

Anyone know where to find the source code?

**Comment by halgab** [comment]
You can't find it because it's not in the main branch yet. See issue #8790. PR #9560 will add it to .NET 9 though.

**Reply by ulfemsoy** [reply]
Thank you for your reply. 
Is it somehow possible to get access to the code of TabControlDesigner prior to .NET 9?

**Reply by halgab** [reply]
I'm not sure what you're asking for. The OOP TabControlDesigner is not supported prior to .NET 9. If you're looking at the .Net Framework version or the VS versions, my understanding (I'm not a Microsoft employee) is that it's closed source. But the code in #9560 should be close enough.

**Reply by ulfemsoy** [reply]
Thank you very much.

I might be misunderstanding the different versions of .NET. 

I recently converted a .NET Framework 4.8 solution to .NET 6 using the "Upgrade Assistant" plug-in. After the conversion, all my customised ControlDesigners and UITypeEditors stopped working. I assumed that this was because OOP ControlDesigners are now required.

The source code of the TabControlDesigner is available online, but I guess it won't work for .NET 6?
https://referencesource.microsoft.com/#System.Design/System/Windows/Forms/Design/TabControlDesigner.cs

I might be looking in the wrong place, but I can't find TabControlDesigner in the "feature/9.0" branch or any of the other branches.

**Reply by halgab** [reply]
[This](https://github.com/SimonZhao888/v-weidzh-winforms/blob/Issue_8790_Add_TablelayoutPanelDesignerAndTabControlDesigner/src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/TabControlDesigner.cs) is the code from the unmerged feature branch, but I have no idea how you could use it in .NET 6. Note that I'm no specialist though.

**Reply by ulfemsoy** [reply]
For other developers: 
The issue with custom WinForm designers in .NET 6 was easily resolved by installing the NuGet package **Microsoft.WinForms.Designer.SDK**, and inheriting from the appropriate subclass in the Microsoft.DotNet.DesignTools.Designers namespace instead of System.ComponentModel.Design.

For more information, please refer to:
https://devblogs.microsoft.com/dotnet/custom-controls-for-winforms-out-of-process-designer/

**Comment by KlausLoeffelmann** [comment, marked as answer]
@ulfemsoy: The Designers for the out-of-process WinForms Designer are not part of the .NET (WinForms) Runtime. They belong to Visual Studio and are closed source. They are different from the Designers which we _have_ published as part of the WinForms runtime, in that they are for in-proc scenarios .NET (Core), and therefore not compatible with Visual Studio (which is based on .NET Framework).
